### ISSUE TITLE
Sync producer - Possibility of msg_opaque going out of scope

### ISSUE THREAD

**Issue by jananee** [edenhill/librdkafka#297]
Hi,

   I provided a sync interface to the producer using the solutions you have provided in other issues. I use the msg_opaque to signal the delivery of the current message. Looks like the delivery callback is not called when all brokers are down; only the event callback is called which is fine. Since the event callback is for global events and I want some way to indicate that the current message was not produced within a configured time, I poll for the msg_opaque to change with a time out. As I understand, calling poll might trigger delivery callbacks associated with other messages too. What happens when I produced the first message and all brokers were down and then they come back up and a second message is produced. When I call poll now, will the delivery callback of the first message be called? The msg_opaque associated with the first message would have gone out of scope as it is a local variable. 

Thanks

**Comment by jananee** [comment]
Adding the code below in case you might need it

enum DeliveryStatus {
    SUCCESS = 0,
    FAILED = -1,
    UNKNOWN = 1
};

class MessageDeliveryReportCb : public RdKafka::DeliveryReportCb {
public:
    void dr_cb (RdKafka::Message &message) {
        std::cout << "Message delivery status for message of size " << message.len() << " bytes: " <<
        message.errstr() << std::endl;
        DeliveryStatus *deliveryStatus = (DeliveryStatus*)message.msg_opaque();
        *deliveryStatus = message.err() == RdKafka::ErrorCode::ERR_NO_ERROR ? DeliveryStatus::SUCCESS : DeliveryStatus::FAILED;
    }
};

class MessageEventCb : public RdKafka::EventCb {
public:
    void event_cb (RdKafka::Event &event) {
        switch (event.type())
        {
            case RdKafka::Event::EVENT_ERROR:
                std::cerr << "ERROR (" << RdKafka::err2str(event.err()) << "): " <<
                        event.str() << std::endl;
                break;

            case RdKafka::Event::EVENT_STATS:
                std::cerr << "\"STATS\": " << event.str() << std::endl;
                break;

            case RdKafka::Event::EVENT_LOG:
                fprintf(stderr, "LOG-%i-%s: %s\n",
                        event.severity(), event.fac().c_str(), event.str().c_str());
                break;

            default:
                std::cerr << "EVENT " << event.type() <<
                        " (" << RdKafka::err2str(event.err()) << "): " <<
                        event.str() << std::endl;
                break;
        }
    }
};

MessageEventCb message_event_cb;
MessageDeliveryReportCb message_dr_cb;

status_type KafkaExporter::exportData(const string& message, const string&topicName) {

    status_type status;

    std::string brokers = getenv("kafka_hostname");
    std::string maxMessageSize = getenv("kafka_max_message_size");

    int32_t partition = RdKafka::Topic::PARTITION_UA;

    RdKafka::Conf *conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
    std::string err_str;

    conf->set("message.max.bytes", maxMessageSize, err_str);
    conf->set("compression.codec", "gzip", err_str);

    conf->set("dr_cb", &message_dr_cb, err_str);
    conf->set("event_cb", &message_event_cb, err_str);
    conf->set("metadata.broker.list", brokers, err_str);

    RdKafka::Producer *producer =  RdKafka::Producer::create(conf, err_str);

    if(producer) {

        RdKafka::Conf *tconf = RdKafka::Conf::create(RdKafka::Conf::CONF_TOPIC);
        RdKafka::Topic *topic = RdKafka::Topic::create(producer, topicName, tconf, err_str);

        DeliveryStatus deliveryStatus = DeliveryStatus::UNKNOWN;
        RdKafka::ErrorCode produceStatusCode = producer->produce(topic, partition, RdKafka::Producer::RK_MSG_COPY /* Copy payload */,
                                                    const_cast<char *>(message.c_str()), message.size(), NULL, &deliveryStatus);

        if (produceStatusCode != RdKafka::ERR_NO_ERROR) {
            LOG(INFO) << "% Produce failed: " <<
                    RdKafka::err2str(produceStatusCode) << ", Message size: " << message.length() <<  " bytes" << std::endl;
            status = -1;
        }
        else {
            producer->poll(0);

            long timeElapsed = 0;
            long cutOffInMillis = PRODUCE_CUTOFF_TIME_MS;
            while (deliveryStatus == DeliveryStatus::UNKNOWN && timeElapsed < cutOffInMillis) {
                cout << "Waiting for the message to be delivered..." << endl;
                producer->poll(POLL_TIMEOUT_MS);
                timeElapsed += POLL_TIMEOUT_MS;
            }
            cout << deliveryStatus << endl;
            status = deliveryStatus == DeliveryStatus::SUCCESS ? 0 : -1;
        }
        delete topic;
   }
    else {
        LOG(INFO) << "Failed to create producer: " << err_str << std::endl;
        status = -1;
    }

    delete producer;

    return status;
}

**Comment by edenhill** [comment]
You typically should not create an RdKafka instance for each message you send, that will be very slow (needs to set up broker connections, etc, for each message).
Instead create the RdKafka instance (producer in this case) in your startup routine and then reuse it for each send. Topics should preferably be created early too, or at least cached.

**Comment by edenhill** [comment]
The delivery report callback should be called even if all brokers are down, but it will only be called when the message times out (`message.timeout.ms`) which by default is 5 mintutes.
If your cutOffInMillis is less than 5 minutes and you then destroy the producer, then the message will be lost.

So make sure your `message.timeout.ms` is in sync with the cutOff timeout.
In fact, you can solely rely on `message.timeout.ms` and simply wait for the deliveryStatus to change.
This will happen within `message.timeout.ms`, either by succesful delivery, or by timeout.

Also make sure to read this:
https://github.com/edenhill/librdkafka/wiki/Proper-termination-sequence

**Comment by jananee** [comment]
Thanks Magnus. Realized that I was not setting the message.timeout.ms right after I posted. I have removed the cutoff timeout, rely only on the message timeout now and It works as expected. I have also made changes to reuse the producer. In my case, the topic is created before the first produce request and the producer will receive produce requests for messages to different topics. Did you mean to say that we need to cache the topic handles? Will it help in my scenario?
